### PR TITLE
Push nested routes

### PR DESCRIPTION
**Pull request by devoncarew**
Some work towards https://github.com/flutter/flutter/issues/4702.

When I run this against flutter_gallery with the route `/buttons`, I get the following exception (on pushing `/buttons`, after pushing `/`):

```
I/flutter : ══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
I/flutter : The following assertion was thrown building Title("Flutter Gallery"; color: Color(0xff009688)):
I/flutter : Unable to find the BuildContext for the Navigator's overlay.
I/flutter : Did you remember to pass the settings object to the route's constructor in your onGenerateRoute
I/flutter : callback?
I/flutter : When the exception was thrown, this was the stack:
I/flutter : #0      ModalRoute.didPush.<anonymous closure> (package:flutter/src/widgets/routes.dart:531)
I/flutter : #1      _AssertionError._checkAssertion (dart:core-patch/errors_patch.dart:31)
I/flutter : #2      ModalRoute.didPush (package:flutter/src/widgets/routes.dart:529)
I/flutter : #3      NavigatorState._push.<anonymous closure> (package:flutter/src/widgets/navigator.dart:401)
I/flutter : #4      State.setState (package:flutter/src/widgets/framework.dart:781)
I/flutter : #5      NavigatorState._push (package:flutter/src/widgets/navigator.dart:396)
I/flutter : #6      NavigatorState.initState (package:flutter/src/widgets/navigator.dart:332)
```

@Hixie, does anything jump out at you here? Something simple? Perhaps something I'm doing that's not expected by the framework?

### CODE REVIEW

**Comment by Hixie** [comment]
@abarth this is hitting your recent work on moving focus. The problem is that we are pushing multiple routes on startup, and so only one is marked as "initial", so we're going down the non-initial codepath, which is expecting to already have been built at least once.

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 1]
```diff
@@ -322,10 +322,16 @@ class NavigatorState extends State<Navigator> {
...
-    )));
+
+    bool isFirst = true;
+    List<String> routes = _splitRoute(config.initialRoute ?? Navigator.defaultRouteName).toList();
+    for (String route in routes) {
+      _push(config.onGenerateRoute(new RouteSettings(
+        name: route,
+        isInitialRoute: isFirst
+      )));
+      isFirst = false;
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/navigator.dart]: What happens if you pass true for `isInitialRoute` for all of these?  You don't want them to animate in either, which means you want them to all be "initial".  (We should probably rename isInitialRoute to something that makes sense in this context).

**Comment by devoncarew** [comment]
It gets a bit further along - the app starts up and displays the main screen, but I get an exception later on - perhaps when responding to the `/buttons` route push?

```
I/flutter : ══╡ EXCEPTION CAUGHT BY SCHEDULER LIBRARY ╞═════════════════════════════════════════════════════════
I/flutter : The following NoSuchMethodError was thrown during a scheduler callback:
I/flutter : The null object does not have a method 'visitChildElements'.
I/flutter : NoSuchMethodError: method not found: 'visitChildElements'
I/flutter : Receiver: null
I/flutter : Arguments: [Closure: (Element) => void]
I/flutter : When the exception was thrown, this was the stack:
I/flutter : #0      Object._noSuchMethod (dart:core-patch/object_patch.dart:42)
I/flutter : #1      Object.noSuchMethod (dart:core-patch/object_patch.dart:45)
I/flutter : #2      Hero.of (package:flutter/src/widgets/heroes.dart:131)
I/flutter : #3      HeroController._updateQuest (package:flutter/src/widgets/heroes.dart:496)
I/flutter : #4      BindingBase&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:476)
I/flutter : #5      BindingBase&SchedulerBinding.handleBeginFrame (package:flutter/src/scheduler/binding.dart:444)
I/flutter : #6      _beginFrame (file:///b/build/slave/Linux_Engine/build/src/out/android_debug_x64/gen/sky/bindings/dart_ui/hooks.dart:48)
I/flutter : ════════════════════════════════════════════════════════════════════════════════════════════════════
```

**Comment by devoncarew** [comment]
If I insert a delay before pushing routes (apart from the first segment), I don't see the exception. Something like:

```
Timer.run(() {
      _push(config.onGenerateRoute(new RouteSettings(
        name: routeFragment,
        isInitialRoute: true
      )));        
    });
```

**Comment by abarth** [comment]
I'm not sure we want to use asynchrony here.  Can you explain a bit more about what you're trying to achieve?  You'd like to start the app up with some history already in place?

**Comment by abarth** [comment]
I might have fixed the hero quest issue in https://github.com/flutter/flutter/pull/4769

**Comment by devoncarew** [comment]
> I'm not sure we want to use asynchrony here. Can you explain a bit more about what you're trying to achieve? You'd like to start the app up with some history already in place?

Yup, if the initial route is `/foo/bar`, I'd like the history to be `/` and `/foo`, and the current page to be `/foo/bar`. When I tried pushing the routes on all at once I saw the exception above. I also saw a separate exception when I tried populating the `_history` list.

I don't think the asynchrony is a _good_ solution here, more me experimenting and trying to get things to work. I'll try a simpler solution w/ your hero fix.

**Pull request closed without merging**